aten/src/ATen/VmapTransforms.cpp

```diff
@@ -94,33 +94,112 @@ static BatchDims computeFrontBatchDimsFromLevels(std::bitset<kVmapNumLevels> lev
+// bitset of vmap levels that were present in the tensor.
+static std::pair<Tensor,std::bitset<kVmapNumLevels>>
+getPhysicalTensorAndLevels(const Tensor& self) {
+  auto* batched = maybeGetBatched(self);
+  if (batched) {
+    return {permuteBatchDimsToFront(batched), createLevelsBitset(batched->bdims())};
   }
-  if (reference_sizes.size() < num_batch_dims) {
-    return false;
+  return {self, 0};
+}
+
+// Given a Tensor or a BatchedTensor, creates a physical view of the tensor
+// such that it has a batch dimension for each levels in `requested_levels`
+// and `requested_example_dim` number of non-batch-dimensions.
+static Tensor alignBatchDimsAtFront(
+    const Tensor& self,
+    std::bitset<kVmapNumLevels> requested_levels,
+    int64_t requested_example_dim) {
+  Tensor physical_tensor;
+  std::bitset<kVmapNumLevels> tensor_levels;
+  std::tie(physical_tensor, tensor_levels) = getPhysicalTensorAndLevels(self);
+
+  TORCH_INTERNAL_ASSERT(
+    (tensor_levels | requested_levels) == requested_levels,
+    "`requested_levels` must be a superset of `self`'s levels");
+
+  auto physical_sizes = physical_tensor.sizes();
+
+  auto tensor_example_dim = physical_sizes.size() - /*num_batch_dims*/tensor_levels.count();
+  TORCH_INTERNAL_ASSERT(tensor_example_dim <= requested_example_dim);
+
+  std::vector<int64_t> aligned_sizes(requested_levels.count() + requested_example_dim, 1);
+
+  // align the example dims (non-bdims dims) first
+  // aligned_sizes[-tensor_example_dim:] = tensor_sizes[-tensor_example_dim:]
+  std::copy(
+      physical_sizes.rbegin(),
+      physical_sizes.rbegin() + tensor_example_dim,
+      aligned_sizes.rbegin());
+
+  // align the bdims
+  int64_t level = 0;
+  int64_t dim = 0;
+  int64_t tensor_dim = 0;
+  for (int64_t level = 0; level < kVmapNumLevels; level++) {
+    if (!requested_levels[level]) {
+      continue;
+    }
+    if (tensor_levels[level]) {
+      aligned_sizes[dim] = physical_sizes[tensor_dim++];
+    }
+    dim++;
   }
-  for (int64_t dim = 0; dim < num_batch_dims; dim++) {
-    if (physical_sizes[dim] != reference_sizes[dim]) {
-      return false;
+
+  return physical_tensor.view(aligned_sizes);
+}
+
+static std::pair<std::bitset<kVmapNumLevels>,int64_t>
+getLevelsAndLargestExampleDim(TensorList logical_tensors) {
+  TORCH_INTERNAL_ASSERT(logical_tensors.size() > 0);
+  std::bitset<kVmapNumLevels> levels;
+  int64_t largest_example_dim = -1;
+  for (const auto& tensor : logical_tensors) {
+    auto* batched = maybeGetBatched(tensor);
+    if (batched) {
+      levels = levels | createLevelsBitset(batched->bdims());
+    }
+    auto ndim = tensor.dim();
+    if (ndim > largest_example_dim) {
+      largest_example_dim = ndim;
     }
   }
-  return true;
+  return { levels, largest_example_dim };
 }
 
-Tensor VmapPhysicalView::newLogicalFromPhysical(const Tensor& physical) {
+std::vector<VmapPhysicalView> BroadcastingVmapTransform::logicalToPhysical(TensorList logical_tensors) {
```

Comment:
Allocation of a vector here is irritating. Maybe a SmallVector here would be better.

Reply:
If performance matters, then I'm planning on having an overload that takes in two tensors and returns a pair of tensors. The most input arguments a broadcasting operators in pytorch is 3 I think, so we could just have an overload for two tensors and one for three tensors.

I'll change this to smallvector for now though